core/src/main/java/me/lokka30/treasury/plugin/core/command/subcommand/economy/migrate/AccountMigrator.java

```diff
@@ -5,161 +5,137 @@
 import me.lokka30.treasury.api.economy.account.Account;
 import me.lokka30.treasury.api.economy.currency.Currency;
-import me.lokka30.treasury.api.economy.transaction.EconomyTransaction;
 import me.lokka30.treasury.api.economy.transaction.EconomyTransactionImportance;
 import me.lokka30.treasury.api.economy.transaction.EconomyTransactionInitiator;
-import me.lokka30.treasury.api.economy.transaction.EconomyTransactionType;
 import org.jetbrains.annotations.NotNull;
 
-// FIXME: Jikoo
 interface AccountMigrator<T extends Account> {
 
-    /*
-    @NotNull String getBulkFailLog(@NotNull Throwable throwable);
+    @NotNull String getBulkFailLog(@NotNull FailureReason failureReason);
 
     @NotNull String getInitLog(@NotNull String identifier);
 
-    @NotNull String getErrorLog(@NotNull String identifier, @NotNull Throwable throwable);
+    @NotNull String getErrorLog(@NotNull String identifier, @NotNull FailureReason failureReason);
 
-    @NotNull BiConsumer<@NotNull EconomyProvider, @NotNull EconomySubscriber<Collection<String>>> requestAccountIds();
+    @NotNull CompletableFuture<Response<Collection<String>>> requestAccountIds(@NotNull EconomyProvider provider);
 
-    @NotNull TriConsumer<@NotNull EconomyProvider, @NotNull String, @NotNull EconomySubscriber<T>> requestAccount();
+    @NotNull CompletableFuture<Response<T>> requestOrCreateAccount(
+            @NotNull EconomyProvider provider, String identifier
+    );
 
-    @NotNull TriConsumer<@NotNull EconomyProvider, @NotNull String, @NotNull EconomySubscriber<Boolean>> checkAccountExistence();
-
-    @NotNull TriConsumer<@NotNull EconomyProvider, @NotNull String, @NotNull EconomySubscriber<T>> createAccount();
-
-    default void migrate(
+    @NotNull
+    default CountDownLatch migrate(
             @NotNull EconomyTransactionInitiator<?> initiator,
-            @NotNull Phaser phaser,
             @NotNull T fromAccount,
             @NotNull T toAccount,
             @NotNull MigrationData migration
     ) {
-        CompletableFuture<Collection<String>> fromCurrencies = new CompletableFuture<>();
-
-        fromAccount.retrieveHeldCurrencies(new PhasedSubscriber<Collection<String>>(phaser) {
-            @Override
-            public void phaseAccept(@NotNull final Collection<String> currencies) {
-                fromCurrencies.complete(currencies);
-            }
-
-            @Override
-            public void phaseFail(@NotNull final EconomyException exception) {
-                migration.debug(() -> getErrorLog(fromAccount.getIdentifier(), exception));
-                fromCurrencies.completeExceptionally(exception);
+        CountDownLatch latch = new CountDownLatch(1);
+        String accountId = fromAccount.getIdentifier();
+        fromAccount.retrieveHeldCurrencies().thenAccept(resp -> {
+            if (!resp.isSuccessful()) {
+                migration.debug(() -> getErrorLog(accountId, resp.getFailureReason()));
+                return;
```

Comment:
Doesn't this result in migration never completing because the latch will never count down? Would prefer to see the migration "complete" with failures rather than leave it stalled indefinitely.

Reply:
see how easy it is to miss such stuff? what's your opinion on migrating that whole latch shit I made to my new library [Process](https://github.com/MrIvanPlays/Process) ?

Reply:
If it makes life easier, all for it. Migration is more than a little messy.